server/tests/bookmark.test.js

```diff
@@ -0,0 +1,322 @@
+import mongoose from 'mongoose';
+import request from 'supertest-as-promised';
+import httpStatus from 'http-status';
+import chai, { expect } from 'chai';
+import app from '../../index';
+import Bookmark from '../models/favorite.model';
+import Post from '../models/post.model';
+import User from '../models/user.model';
+
+chai.config.includeStack = true;
+
+/**
+ * This is duplicate of favorite.test. We can remove favorite once all clients
+ * have migrated to use bookmark routes/elements only
+ */
+
+/**
+ * root level hooks
+ */
+after((done) => {
+  // required because https://github.com/Automattic/mongoose/issues/1251#issuecomment-65793092
+  mongoose.models = {};
+  mongoose.modelSchemas = {};
+  mongoose.connection.close();
+  done();
+});
+
+describe('## Bookmark APIs', () => {
+  const validUserCredentials = {
+    username: 'react',
+    password: 'express'
+  };
+
+  let userToken;
+  let postId;
+  let userId;
+
+  before((done) => {
+    request(app)
+      .post('/api/auth/register')
+      .send(validUserCredentials)
+      .expect(httpStatus.CREATED)
+      .then((res) => {
+        expect(res.body).to.have.property('token');
+        userToken = res.body.token;
+        userId = res.body.user._id;
+        const post = new Post();
+        return post.save();
+      })
+      .then((post) => {
+        postId = post._id;
+        done();
+      })
+      .catch(done);
+  });
+
+  after((done) => {
+    User.remove({}).exec()
+      .then(() => Post.remove({}).exec())
+      .then(() => {
+        done();
+      });
+  });
+
+  afterEach((done) => {
+    Bookmark.remove().exec()
+      .then(() => {
+        done();
+      });
+  });
+
+  describe('# POST /api/posts/:postId/bookmark', () => {
+    it('errors when not logged in', (done) => {
+      request(app)
+        .post(`/api/posts/${postId}/bookmark`)
+        .expect(httpStatus.UNAUTHORIZED)
+        .then((res) => {
+          expect(res.body).to.exist; //eslint-disable-line
+          done();
+        });
+    });
+
+    it('bookmarks a post', (done) => {
+      request(app)
+        .post(`/api/posts/${postId}/bookmark`)
+        .set('Authorization', `Bearer ${userToken}`)
+        .expect(httpStatus.OK)
+        .then((res) => {
+          const bookmark = res.body;
+          expect(bookmark.postId).to.eql(`${postId}`);
+          expect(bookmark.active).to.be.true; //eslint-disable-line
+          expect(bookmark.userId).to.exist; //eslint-disable-line
+          done();
+        })
+        .catch(done);
+    });
+
+    it('toggles the bookmark for a post', (done) => {
+      request(app)
+        .post(`/api/posts/${postId}/bookmark`)
```

Comment:
Ahh, the idea of having this endpoint toggle a bookmark  is something I am not sure about given that we also have unbookmark but  it doesn't matter too much for now. What are your thoughts on that?